### PR TITLE
[Development] Ask for password during encrypted PDF uploads

### PR DESCRIPTION
**Pull request by Mottie**
## Description

Some service members applying for their Benefits Delivery at Discharge will be given a copy of their service treatment record (STR) as an encrypted PDF. There is no easy way to remove this encryption prior to upload. Some backend research discovered it would be possible to do if provided the password. This is the frontend implementation which asks the member to include a password for their encrypted PDF prior to upload.

This PR includes code that auto-detects encrypted PDFs. This includes:
- [TestPDF-password123.pdf](https://github.com/department-of-veterans-affairs/va.gov-team/files/5331074/TestPDF-password123.pdf) (password = `password123`) - Encrypted PDFs that require a password to view
- [TestPDF-edit-password1234.pdf](https://github.com/department-of-veterans-affairs/va.gov-team/files/5331076/TestPDF-edit-password1234.pdf) (password = `password1234`) - Viewable PDFs that require a password to allow editing. Sadly, there is no way to differentiate this from a PDF that requires a password to view
- [TestPDF-encrypted-password123.pdf](https://github.com/department-of-veterans-affairs/va.gov-team/files/5331075/TestPDF-encrypted-password123.pdf) (password = `password123`; but I am unable to include the certificate) - PDF encrypted using a certificate

And an example PDF that doesn't require a password - [TestPDF-no-password.pdf](https://github.com/department-of-veterans-affairs/va.gov-team/files/5331073/TestPDF-no-password.pdf)

Related:
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/14011
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/14002
- https://github.com/department-of-veterans-affairs/vets-api/pull/5086
- https://github.com/department-of-veterans-affairs/vets-api/pull/5072

## Testing done

Updated and added unit tests

## Screenshots

<details><summary>Upload asking for a password</summary>

Screen Shot 2020-10-23 at 1 49 25 PM</details>

<details><summary>Upload after including an invalid password</summary>

Screen Shot 2020-10-23 at 1 44 54 PM</details>

<details><summary>Listing after a valid password has been included</summary>

Screen Shot 2020-10-23 at 1 46 01 PM
</details>

## Acceptance criteria
- [x] Asking for an encrypted PDF password is behind a feature flag (`request_locked_pdf_password`)
- [x] Asking for an encrypted PDF password is behind a uiSchema option (`getEncryptedPassword`)
- [x] Encrypted PDFs are auto-detected and reveal a password input as needed
- [x] Passwords are included with the file upload
- [ ] Backend messaging is displayed to the user as either successful, or unsuccessful

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by rianfowler** [comment]
@department-of-veterans-affairs/vsp-design-system-fe You were not automatically assigned as a required reviewer to this forms system change

**Review by rianfowler — APPROVED** [review]
feature flag change is good- wait for the design system team approval on the form system changes.

**Comment by Mottie** [comment]
Note to reviewer:

- If you hide whitespace changes, the change drops to 799 LOC
  Screen Shot 2020-10-23 at 10 07 48 AM
- Unit tests in the applications add between 20-43 lines of code in 13 files. This is another 355 lines of repetitive code.

So this ends up being about 450 lines of actual code change.

**Review by micahchiang — APPROVED** [review]

File: src/platform/forms-system/src/js/utilities/file/PdfPasswordFeature.js
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/PdfPasswordFeature.js:8]: Not sure why, but using a stateless function here caused the [`getWrappedInstance()`](#diff-be73ef140f22ca1d406a85b603d7b1a3f022dc0ed50550a651c258127b10424cR48) call in `FileField` to show an error.

File: src/platform/forms-system/test/js/utilities/file.unit.spec.js
Comment (bkjohnson) [inline src/platform/forms-system/test/js/utilities/file.unit.spec.js:21]: Why is this mocking happening? I'm a bit nervous about doing this much mocking for a test.

File: src/platform/forms-system/src/js/definitions/file.js
Comment (bkjohnson) [inline src/platform/forms-system/src/js/definitions/file.js:10]: I'm always glad to see lodash go away.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:102]: Since `checkForEncryptedPdf`returns a boolean, can `result` be named to something like `isEncrypted`?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:138]: Can't a lot of this be done with a call to `checkForEncryptedPdf`?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:245]: Is this necessary?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:323]: This seems useful, but it is a pattern that I'm worried could cause unexpected behavior if the rendering takes longer than expected, and it seems like it could be tricky to enforce if we wanted to test that a certain element grabbed focus when we enter an error state. I'm not suggesting anything right now, I'm just calling it out so I can see what others have to say about it.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:352]: We may need approval from a content person on this. I can imagine users being confused when they see `(unlocked)` next to something they uploaded and worrying about that.

Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:352]: We may need approval from a content person on this. I can imagine users being confused when they see `(unlocked)` next to something they uploaded and worrying about that.

Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:352]: We may need approval from a content person on this. I can imagine users being confused when they see `(unlocked)` next to something they uploaded and worrying about that.

**Review by bkjohnson — COMMENTED** [review]
Left some comments so far - I'll probably give it another pass.

File: src/platform/forms-system/src/js/utilities/file/PdfPasswordFeature.js
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/PdfPasswordFeature.js:8]: Because I didn't want to use `connect` in the FileField component, which I did to avoid adding a Provider & store is all the unit tests - which may not be necessary anymore, now that I've ended up doing it anyway - I added this `PdfPasswordFeature` component to get the feature toggle state using `connect`. The component includes a property with the set state.

When the `withRef` is set (see `react-redux` v5.x [`connectAdvanced`](https://react-redux.js.org/5.x/api/connect-advanced#arguments) docs - and honestly, I don't know why `connectAdvanced` showed an error, but a regular `connect` worked) it exposes the component within the `connect` so we can use a reference to target it.

Within the `FileField`, we have to use `getWrappedInstance()` to get access to the wrapped component in order to get the feature flag state.

Honestly, now that I look over this mess, I think I'll refactor this to have the `connect` within the `FileField` and ping you once that's done.

File: src/platform/forms-system/src/js/utilities/file/ShowPdfPassword.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/ShowPdfPassword.jsx:4]: I will try that if the refactor I'm planning doesn't work as I expect.

File: src/platform/forms-system/test/config/helpers.js
Comment (Mottie) [reply src/platform/forms-system/test/config/helpers.js:54]: The `uploadFeatureData` contains the `request_locked_pdf_password` feature flag. I called it `uploadFeatureData` because it was shorter than `requestLockedPdfPasswordFeatureData`. The `data` is in there because it is added to the `data` within the `DefinitionTester`.

File: src/platform/forms-system/test/js/utilities/file.unit.spec.js
Comment (Mottie) [reply src/platform/forms-system/test/js/utilities/file.unit.spec.js:21]: I'm mocking the binary data of an encrypted PDF file. It includes an `/Encrypt` string within the file, and it's the only way I know how to detect if a PDF is encrypted.

I could have used `readAsBinaryString` instead of `readAsArrayBuffer` to simplify the `checkForEncryptedPdf` code, but [we support IE11](https://developer.mozilla.org/en-US/docs/Web/API/FileReader#Browser_compatibility) which just happens to not support `readAsBinaryString`.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:352]: Oops, this is old code that should be removed.... "(unlocked)" isn't even in the design anymore.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:138]: Maybe? These checks are done immediately after the user chooses a file to upload, so checking for a feature flag, existing password, correct file extension and uiSchema setting is much less expensive than calling a Promise. Or maybe I'm wrong? I've refactored this code so many times, I wouldn't be surprised if there was a more efficient method I overlooked.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:245]: So the `formData` saves the server response

```
  /**
   * @typeof Files~ApiResponse
   * @type {object}
   * @property {string} name - file name
   * @property {string} confirmationCode - uuid of uploaded file
   * @property {string} attachmentId - form ID set by user
   */
```

where as the `this.state.files` stores

```
  /**
   * @typedef Files~state - File state object
   * @type {object}
   * @property {object|null} file - upload API supplied File object: see
   *   https://developer.mozilla.org/en-US/docs/Web/API/File
   * @property {number} index - index matching formData array position
   * @property {boolean} isEncrypted - detected encrypted state of a PDF file
   * @property {boolean} isTemporary - indicates state data that needs needs a
   *   placeholder
   */
```

the `getFiles` function needs to add the state object into the form data files for rendering - the `state.files` haven't been uploaded yet,  so they won't be in the form  data. Including a `name` makes it so we don't have to check if there is a difference in the file list that is to be rendered. Also the `getFiles` includes the original `File` object to pass to the `submitPassword` function once the user clicks the Add password button.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:323]: I thought about applying the focus to the element within the `ShowPdfPassword` component when the component mounted, but then it might be stealing focus from another password input, if more than one is open at a time. I don't know a good solution either.

**Comment by Mottie** [comment]
Thanks for the reviews @bkjohnson! I'll. work on a refactor next week... after I've had time to clear my brain of this jumbled mess of code!

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:245]: It is there to keep the render code cleaner...
- Already uploaded files only include a `name` property, but no `file`
- Files that need a password include both a `file` (to provide for upload once a password is added) and `name` to render a file name without needing the render to check what type of entry it is rendering

Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:245]: I am open to suggestions to make this better.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:352]: Removing this label. It's not in the design

**Comment by Mottie** [comment]
@bkjohnson I've updated this PR with your suggestions... I was able to move the feature flag code into the `FileField` file and convert `ShowPdfPassword` into a functional component. If you need to chat about this PR, I would be happy to discuss it with you any time.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:281]: Instead of modifying `hasErrors` which is downstream of the error source, could we make sure that `file.errorMessage` gets included in the `errors`?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:307]: If `file.errorMessage` was included in `errors`, then we wouldn't have to make this confusingly duplicate check

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:390]: Same idea here.

File: src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js
Comment (bkjohnson) [inline src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js:34]: What has the possibility of throwing the exception here? Is it `file` being null, or can the reader throw an exception as it is reading? I ask because I'm wondering if we could get this function to resolve with a consistent true/false, rather than reject with this error object.

**Review by bkjohnson — COMMENTED** [review]
A few more thoughts.

File: src/platform/forms-system/src/js/actions.js
Comment (Mottie) [reply src/platform/forms-system/src/js/actions.js:278]: Maybe not, it was there until the BE was updated... but now it is :D

File: src/platform/forms-system/src/js/utilities/file/ShowPdfPassword.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/ShowPdfPassword.jsx:43]: To add some gap below the password block... based on the design screenshots - https://github.com/department-of-veterans-affairs/va.gov-team/issues/14011#issuecomment-709325581

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:92]: Ok, I'll move that check back into the `FileField` code.

File: src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js:10]: Ok... I thought you wanted the `checkForEncryptedPdf` file to do all the checking. I'll move the checks to the `FileField` code.

File: src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js:13]: You're right! There wasn't any `onloadend` on MDN, so I copy/pasted this somewhere.. didn't think about it.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:339]: I was trying to figure out why the `file.name` wasn't showing up after an attempted upload... because the `action` wasn't including it... and during troubleshooting, I removed the unnecessary span.

File: src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js
Comment (Mottie) [reply src/platform/forms-system/src/js/utilities/file/checkForEncryptedPdf.js:34]: I just tried uploading a 1 byte text file renamed with a PDF extension, and it didn't have an error.. so I guess we can remove the try/catch block.

**Comment by Mottie** [comment]
Recommendations have been applied. Thanks again!

**Review by ndsprinkle — APPROVED** [review]

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:105]: Is `file` guaranteed to not be null here? The JSDoc for `setFileState` says that _that_ `file` arg can be null, and with the removal of the try/catch in `checkForEncryptedPdf`, I think a null `file` will throw an error. Also, when `onAddFile` is called from `onSubmitPassword` it looks like `file` has the possibility to be null.

So if `checkForEncryptedPdf` throws an error, we may end up in the catch block here, which is only supposed to happen if a file is deleted. Are those the same scenario?

I'm also not sold on the `processFile` name. Would `setEncryptedState` be more descriptive?

**Review by bkjohnson — CHANGES_REQUESTED** [review]
I'm requesting changes specifically for the return values inside of the `forEach`.

For future reference, the rebase made this review more difficult and time consuming. It prevented me from replying to previous comments (like the one about the try/catch), and it also made it more difficult to find the commits that were made after my previous review.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:290]: Not really since the errorSchema would return an empty array. We would need to check the length, then fall back to the `file.errorMessage`

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:248]: DOH, you're right!

And sorry about constanting rebasing the PR... I'll stop doing that unless there's a conflict (I don't like to use merge master to update the branch)

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:105]: - The `file` could have been `null` in the `state` data, but I ended up removing all the `state` data and added the pre-uploaded file directly into the `formData`.

The `onAddFile` is wrapped in a function that checks for the `file`, and I've added an additional check that the file includes a `name` that ends in `pdf`. Then `processFile` is called. All of this happens immediately after the user selects a file, so the odds of `file` being `null` at this point would be essentially non-existent. The problem is that `file` may throw an error _after_ the user enters a password and clicks the "Add password" button if the file was deleted in that period of time - in which case `file` _is_ defined, but does not reference an actual file. I just tried this scenario and got a Network error

Screen Shot 2020-10-28 at 12 42 19 PM

Lastly, I'll rename this function to make it more clear

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:81]: It seems like lodash can be avoided here:


```js
    const { onChange, formData = [] } = this.props;
    return checkForEncryptedPdf(file)
      .then(isEncrypted => {
        if (isEncrypted) {
          onChange(() => {
            formData[index] = {
              file,
              name: file.name,
              index,
              isEncrypted: true,
            };
          });
 
```

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:69]: I'm not sold on the name `isFileEncrypted` because this function is doing more than just checking to see if a file is encrypted. It is calling an event handler as well as possibly updating the state.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:141]: Why are we calling `onChange` twice? The `isFileEncrypted` call will call it if the file is encrypted, and once we return true we call it again here.

**Review by bkjohnson — COMMENTED** [review]
Each time I review it something new catches my eye :sweat_smile:

File: src/platform/forms-system/src/js/definitions/file.js
Comment (Mottie) [reply src/platform/forms-system/src/js/definitions/file.js:34]: Not that I can think of, but we're rolling this out for form 526 first, then once we know it's stable, we can set the other forms to add this `uiSchema` flag, or maybe completely remove it along with the feature flag.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:81]: Oh good idea! I was copying how it was done elsewhere in the code. I guess updating them all would be out of scope...

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:69]: I moved all the `onChange` calls out of it... so now it's bare-bones!

```js
  isFileEncrypted = async file =>
    checkForEncryptedPdf(file)
      .then(isEncrypted => isEncrypted)
      // This _should_ only happen if a file is deleted after the user selects
      // it for upload
      .catch(() => false);
```

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:141]: Good catch! Added a single call to `onChange` above the `if (needsPassword) {` check..

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx]: I know this is in a callback, but do we need to call `onChange` here if we are going to call it up above when we check if the user needs to provide a password?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:56]: No, it's related to the password code.

What happens is when a user chooses an encrypted PDF, it adds an entry in `formData` that includes a `file` value (this is the [browser API `File` object](https://developer.mozilla.org/en-US/docs/Web/API/File/File)). But if a user decides to finish the form later, the `File` object becomes an empty object within the save-in-progress data. Upon returning to the page, the empty file object would prevent them from using the same entry to add a password because there is no `File` object pointing to their local device. This code removes the empty `file` object.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:209]: It's not a bug fix, I guess I'm just being extra cautious. Do you want me to remove the fallback?

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx]: This callback `onChange` is called for non-PDF files; the previous `onChange` is exclusive for all PDFs.... I'll move that `onChange` inside the `needsPassword` check so it'll only apply to encrypted PDFs

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (Mottie) [inline src/platform/forms-system/src/js/fields/FileField.jsx:385]: These changes make the `span` the same size as the `label` when it has a focus highlight. It also increases the upload button target size to make it more accessible.

**Comment by Mottie** [comment]
Ready for review... we were hoping to have this merged into staging by tomorrow for sprint review & demo, but I can always do the demo from my local environment if it's not 100% ready.

File: src/platform/forms-system/src/js/definitions/file.js
Comment (Mottie) [reply src/platform/forms-system/src/js/definitions/file.js:34]: I discussed this with the team. Let's start with form 526 as an MVP, then once we've verified it's working, we'll remove the `uiSchema` & feature flag completely. We need both flags to limit usage to form 526 only.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:112]: It may be useful to have a unit test to check that an `onChange` is called in this situation but `uploadFile` is _not_ called.

File: src/platform/forms-system/src/js/definitions/file.js
Comment (bkjohnson) [reply src/platform/forms-system/src/js/definitions/file.js:34]: Just putting this here to document our slack convo - I still think this should be removed.

**Review by bkjohnson — COMMENTED** [review]
A few comments on unit tests, but once those are addressed and `getEncryptedPassword` is removed this should be :+1:

**Review by bkjohnson — APPROVED** [review]
:sweat_smile:

**Pull request merged**
